src/parser.js

```diff
@@ -2245,11 +2311,24 @@ export class GenericParser extends Tokenizer {
     let token = this.lookahead;
     let startState = this.startNode();
 
+    let preAsyncTokenState = this.saveLexerState();
+
+    let isAsync = !!this.eat(TokenType.ASYNC);
+    if (isAsync && this.hasLineTerminatorBeforeNext) {
+      isAsync = false;
+      this.restoreLexerState(preAsyncTokenState);
+    }
+
     let isGenerator = !!this.eat(TokenType.MUL);
+    if (isAsync && !this.lookaheadPropertyName()) {
+      isAsync = false;
+      this.restoreLexerState(preAsyncTokenState);
+      isGenerator = !!this.eat(TokenType.MUL);
```

Comment:
If `isAsync` is true, that means the first token after `preAsyncTokenState` is `ASYNC`, i.e. not `MUL`, so you can just put `isGenerator = false;` here (and for consistency move it up a line).